dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/expression/PredictorPreprocessorExpression.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright (c) 2004-2022, University of Oslo
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ * Redistributions of source code must retain the above copyright notice, this
+ * list of conditions and the following disclaimer.
+ *
+ * Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following disclaimer in the documentation
+ * and/or other materials provided with the distribution.
+ * Neither the name of the HISP project nor the names of its contributors may
+ * be used to endorse or promote products derived from this software without
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.hisp.dhis.expression;
+
+import static java.lang.String.format;
+
+import lombok.Getter;
+
+import org.hisp.dhis.antlr.ParserException;
+
+/**
+ * Parser for predictor expression preprocessing.
+ *
+ * @author Jim Grace
+ */
+@Getter
+public class PredictorPreprocessorExpression
+    extends PreprocessorExpression
+{
+    String variable;
+
+    String taggedDegUid;
+
+    String degUid;
+
+    /**
+     * Accepts and parses a predictor expression with a preprocessor prefix:
+     * forEach ?variable in :DEG:degUid -->
+     *
+     * @param expression the expression to parse for preprocessing.
+     */
+    public PredictorPreprocessorExpression( String expression )
+    {
+        super( expression );
+
+        if ( parts.size() != 4 )
+        {
+            throw new ParserException( format( "Predictor preprocessor expression should have four parts: '%s'",
```

Comment:
Do we want to use an `ErrorCode` for these exceptions to allow programmatic handling on the client-side?

Reply:
Yes, that would be nice. As it stands now, all the parsing uses `ParserException` including many parsing errors that are generated from the [dhis2-antlr-expression-parser](https://github.com/dhis2/dhis2-antlr-expression-parser) repo that was written before the backend error codes were invented and doesn't have access to the backend error codes anyway. I'm hoping that when @jbee and I integrate the PEG parser into core we can find a more clever way to have error codes throughout the parsing errors including from another repo, something like a namespace or numeric range perhaps.

It's even worse than that. All the user sees in the Maintenance app for a bad expression is "Expression is not well-formed". They never see even these messages. When the ANTLR parsing code was released there wasn't even room in the WebAPI for an expression error message to be returned to the frontend. There seems to be now but it's always just the text for the `ExpressionValidationOutcome` enum and the only values used are `VALID` and `EXPRESSION_IS_NOT_WELL_FORMED`. We will need to overhaul this as well when the PEG parser is integrated so we can have meaningful code and meaningful text both go across the WebAPI.

I don't think it makes sense at this point to have these parsing exceptions different from the others; we need to overhaul them all.